core/src/desul/atomics/SYCL.hpp

```diff
@@ -9,135 +9,48 @@ SPDX-License-Identifier: (BSD-3-Clause)
-}
-
-// Atomic Dec
-template<class MemoryOrder/*, class MemoryScope*/>
-inline
-unsigned int atomic_fetch_dec(unsigned int* dest, unsigned int val, MemoryOrder memory_order, MemoryScopeDevice memory_scope) {
-  return atomic_fetch_sub(dest, val, memory_order, memory_scope);
-}
-
-// Atomic Max
-template<class T, class MemoryOrder/*, class MemoryScope*/>
-inline
-typename std::enable_if<Impl::is_sycl_atomic_type<T>::value,T>::type
-atomic_fetch_max(T* dest, T val, MemoryOrder, MemoryScopeDevice) {
-  DESUL_SYCL_NAMESPACE::atomic_ref<
-    T,
-    DesulToSYCLMemoryOrder<MemoryOrder>::value,
-    DesulToSYCLMemoryScope<MemoryScopeDevice>::value,
-    sycl::access::address_space::global_device_space>
-  dest_ref(*dest);
-  return dest_ref.fetch_max(val);
-}
-
-// Atomic Min
-template<class T, class MemoryOrder/*, class MemoryScope*/>
-inline
-typename std::enable_if<Impl::is_sycl_atomic_type<T>::value,T>::type
-atomic_fetch_min(T* dest, T val, MemoryOrder, MemoryScopeDevice) {
-  DESUL_SYCL_NAMESPACE::atomic_ref<
-    T,
-    DesulToSYCLMemoryOrder<MemoryOrder>::value,
-    DesulToSYCLMemoryScope<MemoryScopeDevice>::value,
-    sycl::access::address_space::global_device_space>
-  dest_ref(*dest);
-  return dest_ref.fetch_min(val);
-}
-
-// Atomic And
-template<class T, class MemoryOrder/*, class MemoryScope*/>
-inline
-typename std::enable_if<Impl::is_sycl_atomic_type<T>::value,T>::type
-atomic_fetch_and(T* dest, T val, MemoryOrder, MemoryScopeDevice) {
-  DESUL_SYCL_NAMESPACE::atomic_ref<
-    T,
-    DesulToSYCLMemoryOrder<MemoryOrder>::value,
-    DesulToSYCLMemoryScope<MemoryScopeDevice>::value,
-    sycl::access::address_space::global_device_space>
-  dest_ref(*dest);
-  return dest_ref.fetch_and(val);
-}
-
-// Atomic XOR
-template<class T, class MemoryOrder/*, class MemoryScope*/>
-inline
-typename std::enable_if<Impl::is_sycl_atomic_type<T>::value,T>::type
-atomic_fetch_xor(T* dest, T val, MemoryOrder, MemoryScopeDevice) {
-  DESUL_SYCL_NAMESPACE::atomic_ref<
-    T,
-    DesulToSYCLMemoryOrder<MemoryOrder>::value,
-    DesulToSYCLMemoryScope<MemoryScopeDevice>::value,
-    sycl::access::address_space::global_device_space>
-  dest_ref(*dest);
-  return dest_ref.fetch_xor(val);
-}
 
-// Atomic OR
-template<class T, class MemoryOrder/*, class MemoryScope*/>
-inline
-typename std::enable_if<Impl::is_sycl_atomic_type<T>::value,T>::type
-atomic_fetch_or(T* dest, T val, MemoryOrder, MemoryScopeDevice) {
-  DESUL_SYCL_NAMESPACE::atomic_ref<
-    T,
-    DesulToSYCLMemoryOrder<MemoryOrder>::value,
-    DesulToSYCLMemoryScope<MemoryScopeDevice>::value,
-    sycl::access::address_space::global_device_space>
-  dest_ref(*dest);
-  return dest_ref.fetch_or(val);
-}
+#define DESUL_IMPL_SYCL_ATOMIC_FETCH_OPER(OPER, TYPE)                        \
+  template <class MemoryOrder, class MemoryScope>                            \
+  TYPE atomic_fetch_##OPER(TYPE* dest, TYPE val, MemoryOrder, MemoryScope) { \
```

Comment:
Make sure you see MemoryScope is unconstrained here

Reply:
Since we are using `MemoryScope::Device` and `MemoryScope::Core` for `Cuda` and `HIP`, I think we should do the same and use these overloads. On the other hand, we are not quite sure if the implementation (including the mapping to `address_scope` works correctly for `System` and `Node`) so I would be cautious and delay allowing them. I think it makes sense to use it unrestricted in the end but at the moment the compiler implementation doesn't follow the standard very well and doesn't even define generic address scope. I would wait at least until then.

Reply:
Done